### PR TITLE
[tune] Add external hooks in WandbLoggerCallback

### PR DESCRIPTION
**Pull request by nikitavemuri**
## Why are these changes needed?
This is an experimental feature, so the following changes are added only to the `WandbLoggerCallback`. We are planning to collect feedback about usage and accordingly update or add these changes to the other W&B integration interfaces.
- Allow reading the W&B project name and group name from environment variable if not already passed to callback
- Add external hooks to fetch W&B API key, and to process any information about W&B run

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/air/callbacks/wandb.py [hunk 1]
```diff
@@ -87,6 +92,14 @@ def _set_api_key(api_key_file: Optional[str] = None, api_key: Optional[str] = No
             raise ValueError("Both WandB `api_key_file` and `api_key` set.")
         with open(api_key_file, "rt") as fp:
             api_key = fp.readline().strip()
+    # Try to get API key from external hook
+    if not api_key and WANDB_SETUP_API_KEY_HOOK in os.environ:
+        try:
+            api_key = _load_class(os.environ[WANDB_SETUP_API_KEY_HOOK])()
+        except Exception:
+            logger.exception(
+                f"Error executing {WANDB_SETUP_API_KEY_HOOK} to " "setup API key."
```
Comment (xwjiang2010) [inline python/ray/air/callbacks/wandb.py:101]: wrong format?

File: python/ray/_private/test_utils.py [hunk 2]
```diff
@@ -1366,3 +1366,11 @@ def job_hook(**kwargs):
     cmd = " ".join(kwargs["entrypoint"])
     print(f"hook intercepted: {cmd}")
     sys.exit(0)
+
+
+def wandb_setup_api_key_hook():
```
Comment (xwjiang2010) [inline python/ray/_private/test_utils.py:1371]: hmm, is it used in other tests? Or should we just move it to the tune test below?

File: python/ray/air/callbacks/wandb.py [hunk 3]
```diff
@@ -148,7 +161,14 @@ def __init__(
     def run(self):
         # Since we're running in a separate process already, use threads.
         os.environ["WANDB_START_METHOD"] = "thread"
-        wandb.init(*self.args, **self.kwargs)
+        run = wandb.init(*self.args, **self.kwargs)
+
+        # Run external hook to process information about wandb run
+        if WANDB_PROCESS_RUN_INFO_HOOK in os.environ:
```
Comment (xwjiang2010) [inline python/ray/air/callbacks/wandb.py:167]: do we have test for this code path?

File: python/ray/air/callbacks/wandb.py [hunk 4]
```diff
@@ -9,6 +9,7 @@
 import urllib
 
 from ray import logger
+from ray._private.storage import _load_class
```
Comment (xwjiang2010) [inline python/ray/air/callbacks/wandb.py:12]: is it supposed to be used by libraries? Seems private interface?

**Comment by xwjiang2010** [comment]
left a few comments. Also cc @krfricke for review as well.

File: python/ray/air/callbacks/wandb.py (hunk 3 above)
Comment (nikitavemuri) [reply python/ray/air/callbacks/wandb.py:175]: I added a unit test for this that only calls `_WandbLoggingProcess.run` by mocking out calls to `wandb`

File: python/ray/air/callbacks/wandb.py [hunk 5]
```diff
@@ -87,6 +100,14 @@ def _set_api_key(api_key_file: Optional[str] = None, api_key: Optional[str] = No
...
         with open(api_key_file, "rt") as fp:
             api_key = fp.readline().strip()
+    # Try to get API key from external hook
+    if not api_key and WANDB_SETUP_API_KEY_HOOK in os.environ:
+        try:
+            api_key = _load_class(os.environ[WANDB_SETUP_API_KEY_HOOK])()
+        except Exception:
+            logger.exception(
+                f"Error executing {WANDB_SETUP_API_KEY_HOOK} to setup API key."
+            )
```
Comment (krfricke) [inline python/ray/air/callbacks/wandb.py:110]: Can we log the exception here?
```suggestion
        except Exception as e:
            logger.exception(
                f"Error executing {WANDB_SETUP_API_KEY_HOOK} to setup API key: {e}",
                exc_info=e
            )
```

File: python/ray/air/callbacks/wandb.py [hunk 6]
```diff
@@ -148,7 +169,14 @@ def __init__(
...
         os.environ["WANDB_START_METHOD"] = "thread"
-        wandb.init(*self.args, **self.kwargs)
+        run = wandb.init(*self.args, **self.kwargs)
+
+        # Run external hook to process information about wandb run
+        if WANDB_PROCESS_RUN_INFO_HOOK in os.environ:
+            try:
+                _load_class(os.environ[WANDB_PROCESS_RUN_INFO_HOOK])(run)
+            except Exception:
+                logger.exception(f"Error calling {WANDB_PROCESS_RUN_INFO_HOOK}.")
```
Comment (krfricke) [inline python/ray/air/callbacks/wandb.py:179]: Same here

```suggestion
            except Exception as e:
                logger.exception(f
                    "Error calling {WANDB_PROCESS_RUN_INFO_HOOK}: {e}",
                    exc_info=e
                )
```

**Review by krfricke — COMMENTED** [review]
LGTM, let's just print the exceptions

File: python/ray/air/callbacks/wandb.py (hunk 5 above)
Comment (nikitavemuri) [reply python/ray/air/callbacks/wandb.py]: sure, sounds good

**Review by krfricke — APPROVED** [review]
Thanks! WIll merge once CI passes

**Comment by krfricke** [comment]
@nikitavemuri can you sign off your commits? It would be great if you could follow the instructions in the "DCO" test: https://github.com/ray-project/ray/pull/26346/checks?check_run_id=7365417488

**Comment by nikitavemuri** [comment]
@krfricke I wasn't able to sign off the previous commits in this branch, so I created a new PR at https://github.com/ray-project/ray/pull/26617. Could you take a look at that one?

**Pull request closed without merging**